src/nix-store/nix-store.cc

```diff
@@ -364,6 +365,39 @@ static void opQuery(Strings opFlags, Strings opArgs)
             }
             break;
 
+        case qUnsubstitutable: {
+            if (opArgs.empty()) throw UsageError("no paths passed");
+
+            StorePathSet storePaths;
+            for (auto const & str : opArgs) {
+               storePaths.emplace(store->followLinksToStorePath(str));
+            }
+
+            StringSet pathsS;
+            // query each configured substituter
+            for (auto & sub : getDefaultSubstituters()) {
+              if (sub->storeDir != store->storeDir) continue;
+              if (!sub->wantMassQuery) continue;
+
+              auto paths = sub->queryValidPaths(StorePathSet(storePaths.begin(), storePaths.end()));
+              for (auto & path : paths) {
+                pathsS.emplace(store->printStorePath(path));
+              }
+            }
+
+            StorePathSet substitutablePaths;
+            for (auto & str : pathsS) {
+              substitutablePaths.emplace(store->followLinksToStorePath(str));
+            }
+
+            for ( auto & path : storePaths) {
+                if (substitutablePaths.find(path) == substitutablePaths.end())
```

Comment:
```suggestion
                if (!substitutablePaths.count(path))
```

Reply:
This does not short-circuit

Reply:
In what way? On an `std::map` it should be equally efficient.

But since we're fine with C++20 now, you can also use `substitutablePaths.contains(path)`.